[#43692] Part 1. Modify API to be able to handle file_links without a container.

ba1ash opened the pull request:
https://community.openproject.org/work_packages/43692

The PR adds a new endpoint `POST /api/v3/file_links` and a cron job to delete hung file_links.

ba1ash committed:
[#43692] Modify API to be able to handle file_links without a container.

https://community.openproject.org/work_packages/43692

Kharonus commented on modules/storages/lib/api/v3/file_links/file_link_representer.rb:
```diff
@@ -144,7 +145,7 @@ def _type
 
     def user_allowed_to_manage?(model)
-      current_user.allowed_to?(:manage_file_links, model.container.project)
+      model.container && current_user.allowed_to?(:manage_file_links, model.project)
```
minor: can you make that `nil` check here explicit? With 

```ruby
model.container.present?
```

ba1ash replied:
Yes, sure.

Kharonus commented on modules/storages/spec/requests/api/v3/file_links/file_links_spec.rb:
```diff
@@ -91,6 +91,76 @@
+
+      it 'does not provide a link to the collection of created file links' do
+        self_link = JSON.parse(response.body).dig('_links', 'self', 'href')
+        expect(self_link).to eq('urn:openproject-org:api:v3:file_links:no_link_provided')
```
Use capybara matcher for json pathes:

```ruby
expect(response.body).to be_json_eql('urn:openproject-org:api:v3:file_links:no_link_provided').at_path('_links/self/href')
```

Kharonus replied:
same below

Kharonus commented on modules/storages/spec/requests/api/v3/file_links/file_links_spec.rb:
```diff
@@ -91,6 +91,76 @@
+    let(:path) { '/api/v3/file_links' }
+    let(:permissions) { %i(manage_file_links) }
+    let(:storage_url1) { storage.host }
+    let(:storage_url2) { storage.host }
```
No need to use 2 times the same storage host in different variables, or do I miss something?

ba1ash replied:
Im not sure. I will check it.

Kharonus commented on modules/storages/spec/requests/api/v3/file_links/file_links_spec.rb:
```diff
@@ -91,6 +91,76 @@
+          set_keys.each do |key|
+            expect(file_link.attributes[key]).not_to(
+              be_nil,
+              "expected keyibute #{key.inspect} of FileLink ##{i + 1} to be set.\ngot nil."
```
what is a keyibute?

ba1ash replied:
It is an orphan of a substitute command run on a region. I will fix it.

Kharonus reviewed (changes requested):
Only some minor issues.

ba1ash committed:
Address review comments.

- Use more explicit present? call to check container existence.
- Improve rspec style.

Kharonus reviewed (approved):
LGTM
